crates/ruff_formatter/src/printer/mod.rs

```diff
@@ -322,8 +327,8 @@ impl<'a> Printer<'a> {
 
         self.print_str(text);
 
-        if source_position.is_some() {
-            self.state.source_position += text.text_len();
+        if let Some(range) = source_range {
+            self.state.source_position = range.end();
```

Comment:
I'm fairly certain that `+= text.text_len()` produces incorrect results for dynamic texts (text that haven't been present in the source code. 

Let's assume that the formatter replaces `"String with useless \' escape"` with "String with useless ' escape". It will now map the source position of the `"` to after the end of the string, which is incorrect. The result is worse if the new text is longer than the original text, because it then  maps the source position past the end of the source-string.